### PR TITLE
refactor: rewrite TestRun to typescript

### PR DESCRIPTION
**Pull request by miherlosev**
Changes:
* rewrite `TestRun` to TypeScript
* move the `testClient` function and test client settings from the `Gulpfile.js`
* more strict types

### CODE REVIEW

File: src/reporter/index.ts [hunk 1]
```diff
@@ -267,7 +267,10 @@ export default class Reporter {
                 id:   testRun.test.fixture.id
             },
             command: formatCommand(command, result),
-            browser: testRun.controller.browser,
+            // TestController doesn't have the browser property in its definition.
+            // It is dynamically added later.
+            // @ts-ignore
```
Comment (AndreyBelym) [inline src/reporter/index.ts:272]: IMHO, this property should be declared in the TestController interface, even if it is not assigned in the constructor, but some time later in the external code.

File: src/role/role.ts [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    public redirectUrl: string | null;
+    private readonly _initFn: Function | null;
+    public opts: RoleOptions;
+    public initErr: null | Error;
+    public stateSnapshot: StateSnapshot;
+
+    public constructor (loginUrl: string | null, initFn: Function | null, options = {}) {
+        super();
+
+        // @ts-ignore
```
Comment (AndreyBelym) [inline src/role/role.ts:22]: Let's try to declare this property too.

Comment (AndreyBelym) [reply src/role/role.ts:22]: It's possible:
```
const sym: unique symbol = Symbol();

class Sym {
    public [sym]: number;

    constructor () {
        this[sym] = 42;

    }
}
```

File: src/utils/get-options/get-assertion-timeout.ts [hunk 3]
```diff
@@ -0,0 +1,11 @@
+import AssertionCommand from '../../test-run/commands/assertion';
+import { Dictionary } from '../../configuration/interfaces';
+
+export default function getAssertionTimeout (command: AssertionCommand, options: Dictionary<OptionValue>): number {
+    // @ts-ignore
+    const { timeout: commandTimeout } = command.options;
```
Comment (AndreyBelym) [inline src/utils/get-options/get-assertion-timeout.ts:6]:
```suggestion
    const commandTimeout = command.options?.timeout;
```

**Review by AndreyBelym — COMMENTED** [review]
Good job as always, but I hate @ts-ignore :)

File: src/utils/get-options/get-assertion-timeout.ts (hunk 3 above)
Comment (miherlosev) [reply src/utils/get-options/get-assertion-timeout.ts:6]: Unfortunately, it doesn't allow us to remove `@ts-ignore` comment.
The command properties added [dynamically](https://github.com/DevExpress/testcafe/blob/master/src/test-run/commands/assertion.js#L45).

We can fix it using separate type definition, but it's better to rewrite in more object manner. Command rewriting is a big separate task and it's out of context this PR.

**Review by AndreyBelym — APPROVED** [review]

**Pull request merged**